### PR TITLE
deeplearning4j-modelexport-solr: add ModelTupleStream class

### PR DESCRIPTION
**Pull request by cpoerschke**
A new class implementing the abstract [org.apache.solr.client.solrj.io.stream.TupleStream](https://lucene.apache.org/solr/7_5_0/solr-solrj/org/apache/solr/client/solrj/io/stream/TupleStream.html) class which is part of Apache Solr's [Streaming Expressions](https://lucene.apache.org/solr/guide/7_5/streaming-expressions.html) functionality.

testing done:
* `cd deeplearning4j-modelexport-solr ; mvn test`

### CODE REVIEW

File: deeplearning4j/deeplearning4j-modelexport-solr/src/main/java/org/deeplearning4j/nn/modelexport/solr/ltr/model/ScoringModel.java [hunk 1]
```diff
@@ -106,24 +106,28 @@ public float score(float[] modelFeatureValuesNormalized) {
...
+  }
+
   /**
    * Currently supports {@link MultiLayerNetwork} and {@link ComputationGraph} models.
    * Pull requests to support additional <code>org.deeplearning4j</code> models are welcome.
    */
-  public static float outputScore(Model model, float[] modelFeatureValuesNormalized) {
-
-    final INDArray input = Nd4j.create(modelFeatureValuesNormalized);
+  public static INDArray output(Model model, INDArray input) {
```
Comment (cpoerschke) [inline deeplearning4j/deeplearning4j-modelexport-solr/src/main/java/org/deeplearning4j/nn/modelexport/solr/ltr/model/ScoringModel.java]: The changes here to `ScoringModel.java` are unrelated to this pull request and are intended to 'go away' via the #5428 refactor.

**Comment by AlexDBlack** [comment]
@cpoerschke Just checking up on the status here: I gather this PR still can't be merged until the upstream issue you linked is closed/fixed?

**Comment by cpoerschke** [comment]
@AlexDBlack That's correct, this issue is a work-in-progress as it has a dependency on the mentioned issue. I'll prefix the PR title to reflect that.

**Comment by sshepel** [comment]
@cpoerschke could you please following notation for PRs that are in progress state: `[WIP]`, thanks.

**Comment by AlexDBlack** [comment]
@sshepel @cpoerschke I've edited the title to have this.
FYI this ```[WIP]``` is just a way for us to tell our continuous integration (CI) backend not to run tests/checks on this (until that tag is removed and another commit is pushed up)

**Comment by cpoerschke** [comment]
@sshepel @cpoerschke - thanks for mentioning and explaining about the special `[WIP]` tag. I just opened #5884 to perhaps include that (in some form or other, feel free to edit) in the CONTRIBUTING.md docs - what do you think?

**Comment by sshepel** [comment]
@cpoerschke thanks a lot for your efforts! Looks good to me!

**Comment by cpoerschke** [comment]
this pull request is now refreshed but it's still a work-in-progress:
* has a dependency on (not yet released) Apache Lucene/Solr 7.5.0
* has one small TODO in the code still
* not yet started: i'd like to add an integration test of some sort

File: pom.xml [hunk 2]
```diff
@@ -216,7 +230,7 @@
         <typesafe.config.version>1.3.0</typesafe.config.version>
         <lombok.version>1.18.0</lombok.version>
         <cleartk.version>2.0.0</cleartk.version>
-        <lucene-solr.version>7.4.0</lucene-solr.version>
+        <lucene-solr.version>7.5.0-SNAPSHOT</lucene-solr.version> <!-- TODO: once it's released change this to 7.5.0 for https://issues.apache.org/jira/browse/SOLR-12402 change -->
```
Comment (cpoerschke) [inline pom.xml]: pull request cross-reference: i opened separate #6485 for the 7.4.0 to 7.5.0 upgrade since it involves the one-line pom.xml change here plus a couple of java doc updates elsewhere.

**Comment by cpoerschke** [comment]
status update:
* merged latest master (including the  #6485 7.4.0 to 7.5.0 version upgrade) into this pull request
* have no further changes planned at this time but would welcome general and/or specific feedback
* tentatively removed the [WIP] prefix from the pull request (but locally got weird 'could not resolve dependencies' Maven error, will need to look into that further if it's not just a local issue)

**Comment by cpoerschke** [comment]
Made a little progress and in doing so identified a dependency related blocker, so will add back the [WIP] prefix for clarity.

* Local weird build issues mysteriously went away.
* Latest master merge(s) caused test failures with super helpful `DL4JInvalidConfigException` pointing to exactly what needed fixing.
```
Invalid output layer configuration for layer "layer0": loss function LossMCXENT() expects activations to be in the range 0 to 1 (probabilities) but activation function identity does not bound values to this 0 to 1 range. This indicates a likely invalid network configuration. 
This configuration validation check can be disabled for MultiLayerConfiguration and ComputationGraphConfiguration using validateOutputLayerConfig(false), however this is not recommended.
```
* The integration test then gave a strange 'could not find class' style error and I think this is due to a versions difference i.e. current deeplearning4j metrics.version is 3.1.2 but [Solr 7.5.0 uses 3.2.6](https://github.com/apache/lucene-solr/blob/releases/lucene-solr/7.5.0/lucene/ivy-versions.properties#L61) version.
  * The [cdfb445](https://github.com/deeplearning4j/deeplearning4j/pull/5429/commits/cdfb445c632b9a84e0c24e97d5f5577f6ff7ecae) commit's `alt.metrics.version` is an unmergable hack and the next logical step to explore would appear to be to see if the actual `metrics.version` could be upgraded to 3.2.6 version?

File: pom.xml [hunk 3]
```diff
@@ -252,6 +252,7 @@
         <leveldb.version>1.8</leveldb.version>
         <lz4.version>1.3.0</lz4.version>
         <metrics.version>3.1.2</metrics.version>
+        <alt.metrics.version>3.2.6</alt.metrics.version>
```
Comment (cpoerschke) [inline pom.xml]: This `alt.metrics.version` addition here is (intentionally) an unmergeable hack. #6580 is to potentially upgrade the actual `metrics.version` itself.

Comment (saudet) [reply pom.xml]: Please go ahead and try to update `metrics.version` to 3.2.6.

**Comment by cpoerschke** [comment]
put back to work-in-progress since out-of-date w.r.t. base branch

**Comment by saudet** [comment]
@cpoerschke Do you still need to fix something before we can merge?

**Comment by cpoerschke** [comment]
> @cpoerschke Do you still need to fix something before we can merge?

It would make sense to upgrade to 7.7.1 Apache Lucene/Solr version I think -- opened #7539 for that -- though that would not be a requirement. After the version upgrade the java doc links on the pull request here would need to be edited to match. Other than that no fixes needed as far as I'm aware. Thank you.

**Comment by saudet** [comment]
@cpoerschke I think we just need to add Skymind copyright headers to the new files and we should be good!

**Comment by saudet** [comment]
The copyright could be from some other entity as well I think as long as the license stays the same it shouldn't cause problems.

**Comment by cpoerschke** [comment]
> ... I think we just need to add Skymind copyright headers to the new files and we should be good! ...

Done. And I've also updated the 7.5.0 references to 7.7.1 in the javadocs and the test `solrconfig.xml` file.

**Comment by saudet** [comment]
@cpoerschke The tests are failing though. Can we get this fixed before merging?
```
org.deeplearning4j.exception.DL4JInvalidInputException: Input that is not a matrix; expected matrix (rank 2), got rank 1 array with shape [1]. Missing preprocessor or wrong input type? (layer name: layer0, layer index: 0, layer type: OutputLayer)
	at org.deeplearning4j.nn.modelexport.solr.handler.ModelTupleStreamTest.doTest(ModelTupleStreamTest.java:152)
	at org.deeplearning4j.nn.modelexport.solr.handler.ModelTupleStreamTest.test(ModelTupleStreamTest.java:85)
```
We probably just need to update the tests a bit to create explicitly arrays of rank 2, which is no longer the minimum. ND4J now also supports rank 0 and 1.

**Comment by cpoerschke** [comment]
> ... The tests are failing though. Can we get this fixed before merging? ...

Of course, apologies, I'd seen test failures locally but had hoped/assumed that it's just a localised build issue. Thanks for the pointers for the test failure. I'd like to start by fixing the test failure in the existing code -- #7569 opened for that -- and then will next apply likely the same fix here.

**Comment by saudet** [comment]
@AlexDBlack The tests now pass, when called with `-Dtest.solr.allowed.securerandom=NativePRNG` because of issue #7570. Good enough to merge?
